File: dsa-problems/Tries/solutions/cpp/TRI-011-suffix-trie-longest-repeat.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <algorithm>
using namespace std;

struct TrieNode {
    unordered_map<char, TrieNode*> children;
    int suffixCount = 0;  // Number of suffixes passing through this node
};

class Solution {
private:
    TrieNode* root;
    int maxLength;

    void insertSuffix(const string& suffix) {
        TrieNode* node = root;
        for (char c : suffix) {
            if (node->children.find(c) == node->children.end()) {
                node->children[c] = new TrieNode();
            }
            node = node->children[c];
            node->suffixCount++;  // Increment count for each suffix passing through
        }
    }

    void dfs(TrieNode* node, int depth) {
        // A repeated substring exists if 2+ suffixes pass through this node
        if (node->suffixCount >= 2 && depth > 0) {
            maxLength = max(maxLength, depth);
        }

        for (auto& [ch, child] : node->children) {
            dfs(child, depth + 1);
        }
    }

public:
    Solution() {
        root = new TrieNode();
        maxLength = 0;
    }

    int longestRepeatedSubstring(const string& s) {
        for (int i = 0; i < s.length(); i++) {
            insertSuffix(s.substr(i));
        }

        dfs(root, 0);
        return maxLength;
    }
};

int main() {
    string s;
    getline(cin, s);

    Solution solution;
    int result = solution.longestRepeatedSubstring(s);
    cout << result << endl;

    return 0;
}

File: dsa-problems/Tries/solutions/javascript/TRI-011-suffix-trie-longest-repeat.js
const readline = require("readline");

class TrieNode {
  constructor() {
    this.children = new Map();
    this.suffixCount = 0; // Number of suffixes passing through this node
  }
}

class Solution {
  constructor() {
    this.root = new TrieNode();
    this.maxLength = 0;
  }

  insertSuffix(suffix) {
    let node = this.root;
    for (const char of suffix) {
      if (!node.children.has(char)) {
        node.children.set(char, new TrieNode());
      }
      node = node.children.get(char);
      node.suffixCount++; // Increment count for each suffix passing through
    }
  }

  dfs(node, depth) {
    // A repeated substring exists if 2+ suffixes pass through this node
    if (node.suffixCount >= 2 && depth > 0) {
      this.maxLength = Math.max(this.maxLength, depth);
    }

    for (const child of node.children.values()) {
      this.dfs(child, depth + 1);
    }
  }

  longestRepeatedSubstring(s) {
    for (let i = 0; i < s.length; i++) {
      this.insertSuffix(s.substring(i));
    }

    this.dfs(this.root, 0);
    return this.maxLength;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = "";
rl.on("line", (line) => {
  input = line.trim();
  rl.close();
}).on("close", () => {
  const solution = new Solution();
  const result = solution.longestRepeatedSubstring(input);
  console.log(result);
});
